gta/map.py

```diff
@@ -0,0 +1,119 @@
+# GTA DragonFF - Blender scripts to edit basic GTA formats
+# Copyright (C) 2019  Parik
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+
+#######################################################
+
+# Direct port of OpenRW's LoaderIPL.cpp ( https://github.com/rwengine/openrw/blob/master/rwengine/src/loaders/LoaderIPL.cpp )
+# Seems to only support model INSTances and ZONEs. Though I believe the
+# INSTances alone should give us enough info to import the entire maps
+# of III / VC / SA.
+
+import os 
+from enum import Enum
+
+#######################################################
+class SectionTypes(Enum):
+	INST = 1
+	PICK = 2
+	CULL = 3
+	ZONE = 4
+	NONE = 5
+
+#######################################################
+def loadIPL(filepath):
+
+	section = SectionTypes.NONE
+
+	with open(filepath) as fp:
+
+		line = fp.readline()
+		linesCount = 0
+		while line:
+
+			linesCount += 1
+			line = line.strip()
+			print("Line {}: {}".format(linesCount, line))
+			
+			if len(line) > 0 and line[0] == "#":
+				# Nothing, just a comment
+				pass
+
+			elif line == "end":
+				# Terminating a section
+				section = SectionTypes.NONE
+
+			# Section definers
+			elif section == SectionTypes.NONE:
+				if line == "inst":
+					section = SectionTypes.INST
+				
+				if line == "pick":
+					section = SectionTypes.PICK
+				
+				if line == "cull":
+					section = SectionTypes.CULL
+				
+				if line == "zone":
+					section = SectionTypes.ZONE
+			
+			# Actual data
+			else :
+				if section == SectionTypes.INST:
+
```

Comment:
Another streak of if-else chains here as well. Now as you already use the split options, you can probably create a class or a namedtuple to automatically set these options, (id, model, posX, etc.). 

These classes can then be referenced from the code by using a map keyed by the SectionTypes we set earlier, or better just set the SectionTypes to be equal to the classes/namedtuples which could then have functions like `readSection` to read the entire section. 

Something like this
```python
class Section_INST:
     # properties
     def read(data):
          # put read functions here.

     def write(self):
          # return written data

#...

sections = {
     'inst': Section_INST
}

# then in the readIpl function,
if name in sections:
     section = sections[line].read(data)
```

Reply:
So the read method should be static, read the entire section and return an array of INST objects or ZONE objects etc.?

And what exactly will the write function do?

Reply:
Yeah, well it is not important for it to be a static function, you could probably create an object and call the function on that. 
If it is static though, it should return a new object of that class, so that they can be interacted with later.

It's not important for the write function to be functional or present at the moment, but I would like if you're able to both read, edit and create these files and not just read them. So it could basically act as a Map Editor as well.

Reply:
Oh right get it now. 👍